### DISCUSSION TITLE
Remote Rendering with Paraview Example

### DISCUSSION THREAD

**Discussion by staceywhitmore-inl** [Kitware/trame#531] [Trame/ParaView] [answered]
From the Trame tutorial, how do I combine the venv Paraview example where I execute my script from pvpython using the following command:

`"C:\Program Files\ParaView 5.11.2\bin\pvpython.exe" MyApp.py -venv /c/Path/To/My/Local/miniconda3/envs/trame/
`

with the remote rendering example so I can run two or more independent client processes with trame tools like in this command:
`python -m trame.tools.serve --exec remote_rendering_class:MyClassName
`
?

I've tried combing the syntax from these two commands but to no avail. (E.g., 

`"C:\Program Files\ParaView 5.11.2\bin\pvpython.exe" -m trame.tools.server --exec MyApp:MyClass -venv /c/Path/To/My/Local/miniconda3/envs/trame/`

) but naturally 'pvpython isn't your regular python' as has been mentioned. So, accessing trame.tools.server from pvpython is likely not as simple as combining these two commands on the command line. 

I just want to be able to serve my app that is rendered by Paraview in two different browsers, performing different actions in each connected client session (rather than the default option in which the two client windows will mirror each other bc they use the same instance). Eventually I would like to render this app on a server from a remote Paraview instance but for now I am just running it from localhost.
 
I can get each of these two tutorial examples to work on their own I'm just running into the expected dependency issues when I try to combine the two commands in my CLI.

**Comment by jourdain** [comment, marked as answer]
You might be able to do the following, otherwise you will need to extend PYTHONPATH/PATH

`pvpython.exe -m paraview.apps.trame --trame-app trame.tools.server --venv /c/Path/To/My/Local/miniconda3/envs/trame/ --exec MyApp:MyClass`

**Reply by staceywhitmore-inl** [reply]
I tried this and I still get  this console error:
`ModuleNotFoundError: No module named 'trame.tools.server'`

Will I then need to extend my PYTHONPATH/PATH? If so, am I just adding some items to my PYTHONPATH variable and which paths would I add to it?  

(Also, I should be able to exclude the` :MyClass `part if I'm not using a class in my python file, correct?)

**Reply by jourdain** [reply]
`trame.tools.server` can not work if it is not a class that you can instantiate with a server as an argument.

**Reply by jourdain** [reply]
It seems that your issue is that the venv does not get processed by paraview. Probably because of a mix of path style on Windows?

**Reply by staceywhitmore-inl** [reply]
I am on a windows machine. Are you referring to a using combination of `\` s and `/` or in a path name or an LF/CRLF issue? Perhaps I will try executing the same code on a Linux box?

My venv appears to be working find on other more basic examples. For instance, I can run the aforementioned 
`"C:\Program Files\ParaView 5.11.2\bin\pvpython.exe" MyApp.py -venv /c/Path/To/My/Local/miniconda3/envs/trame/ 
`
just fine. I will try refactoring my python script to wrap everything in an instantiable class and try the command you mentioned again to see if that works...

**Reply by jourdain** [reply]
Thinking more of what you could be doing that I'm not aware of. You might be going down a rabbit hole that could be a dead-end. 

Basically I'm worried that you don't see the implication of using `trame.tools.server` with the ParaView severmanager and that you will not get the multi-user experience that you aim to have. 
While it is possible to do it, it really depend how you organized your ParaView code and what you are truly trying to achieve.

**Reply by staceywhitmore-inl** [reply]
Good point. Ideally I would just look at an example that combines the lessons taught in the [chapter 05_paraview of the Trame tutorial](https://github.com/Kitware/trame-tutorial/tree/master/05_paraview) [SimpleCone ](https://github.com/Kitware/trame-tutorial/blob/master/05_paraview/SimpleCone.py) example, encapsulated in the following command 
`C:\Program Files\ParaView 5.11.2\bin\pvpython.exe" SimpleCone.py --venv C:\Path\To\My\miniconda3\envs\trame\`
with the material taught in the [remote_rendering_class.py example from the Trame course](https://github.com/Kitware/trame-course/blob/master/vtk/remote_rendering_class.py) as executed in the following command
`python -m trame.tools.serve --exec remote_rendering_class:Cone
`
,as simply running the command you mentioned earlier will not work if things are not implemented correctly in the code I'm trying to run, which combines these two concepts. 

Are you aware of such a code example or do you know of a good starting place in the documentation that points to the concepts I will need to learn to make this happen?

I can provide a basic code sample of what I'm trying to do if you think that might be helpful as well.

**Reply by jourdain** [reply]
I guess in the example of chapter 5, we do `view = simple.Render()` which means every application that get started by each user will share the same view where many cones will be added. Then when the session exit, those cones will stay around. So you will keep adding more cones without removing any of them into the same view that will be shared across all sessions.

First try to get the  `remote_rendering_class:Cone` working in the context of ParaView using the `trame.tools.serve` either by using the `paraview.apps.trame` entrypoint or your own that import paraview.web.venv but run trame.tools.serve with your target class...

**Reply by jourdain** [reply]
At that point it is just python issue, but once that is solved, you can start thinking of the specificity of your ParaView app.

**Reply by staceywhitmore-inl** [reply]
OK. I got it to work. Wrapping all my code in a class helped! Thanks for your help! 
`pvpython.exe -m paraview.apps.trame --trame-app trame.tools.serve --venv /c/Path/To/My/Local/miniconda3/envs/trame/ --exec MyApp:MyClass` now works!

Also, I think there may have been a typo in your original answer as it lists `trame.tools.serve**r**` (with an 'r') and `trame.tools.serve` is what worked for me.

**Reply by jourdain** [reply]
indeed it is to `serve` a trame app... https://trame.readthedocs.io/en/latest/tools.serve.html
